org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/StateBasedContainer.java

```diff
@@ -44,19 +46,24 @@ public StateBasedContainer(IResourceDescriptions descriptions, IContainerState s
 	
 	@Override
 	protected Iterable<IEObjectDescription> filterByURI(Iterable<IEObjectDescription> unfiltered) {
-		return Iterables.filter(unfiltered, new Predicate<IEObjectDescription>() {
+		Predicate<IEObjectDescription> predicate = getStateContainsPredicate(input -> input.getEObjectURI().trimFragment());
+		return Iterables.filter(unfiltered, predicate);
+	}
+	
+	private <T> Predicate<T> getStateContainsPredicate(Function<T, URI> uriProvider) {
```

Comment:
am not sure if protected would be better

Reply:
https://github.com/search?q=%22extends+StateBasedContainer+%7B%22&type=Code

Reply:
> i will try to test this next week but am still unsure as we already have extended `EObjectDescriptionLookup` with a type to objects lookup map.
> so we might not really profit from this

we recognized that performance issue on successive pastes in a source file, if there are enough other entries in your index.
you might still recognize the improvement at least in the number of calls on getExportedObjectsByType you get in your EObjectDescriptionLookUp.

Reply:
> protected would be better

:+1: